src/main/kotlin/grpc/AuthenticationInterceptor.kt

```diff
@@ -0,0 +1,298 @@
+
+    val COOKIES_TO_SET_CONTEXT_KEY: Context.Key<MutableMap<String, String?>> = Context.key("cookiesToSet")
+
+    const val ACCESS_TOKEN_COOKIE_NAME = "access_token"
+
+    const val REFRESH_TOKEN_COOKIE_NAME = "refresh_token"
+
+    val COOKIE_METADATA_KEY: Metadata.Key<String> =
+        Metadata.Key.of("cookie", Metadata.ASCII_STRING_MARSHALLER)
+
+    val SET_COOKIE_METADATA_KEY: Metadata.Key<String> =
+        Metadata.Key.of("set-cookie", Metadata.ASCII_STRING_MARSHALLER)
+
+    /**
+     * Retrieves the user ID from the gRPC context.
+     *
+     * @throws SnowballRException.MissingContextException.MissingUserId if the user ID is not in the context.
+     * @return The user ID as a [UUID].
+     */
+    fun getUserIdFromContext(): UUID {
+        val userId = USER_ID_CONTEXT_KEY.get()
+        if (userId == null) {
+            logger.error { "User ID is missing from the gRPC context." }
+            throw SnowballRException.MissingContextException.MissingUserId()
+        }
+
+        return userId
+    }
+
+    /**
+     * Retrieves the cookies to set from the gRPC context.
+     *
+     * @throws SnowballRException.MissingContextException.MissingCookiesMap if the cookies map is not in the context.
+     * @return A mutable map of cookies to set, where keys are cookie names and values are cookie values.
+     */
+    fun getCookiesToSetFromContext(): MutableMap<String, String?> {
+        val cookies = COOKIES_TO_SET_CONTEXT_KEY.get()
+        if (cookies == null) {
+            logger.error { "Cookies map is missing from the gRPC context." }
+            throw SnowballRException.MissingContextException.MissingCookiesMap()
+        }
+
+        return cookies
+    }
+
+    /**
+     * Creates a `Set-Cookie` header string.
+     *
+     * @param name The name of the cookie.
+     * @param value The value of the cookie.
+     * @param maxAgeSeconds The max-age of the cookie in seconds. A value of 0 means delete the cookie.
+     * @param httpOnly If the cookie should be HttpOnly. Defaults to true.
+     * @param secure If the cookie should be Secure. Defaults to true.
+     * @return A formatted `Set-Cookie` string.
+     */
+    fun createCookieString(
+        name: String,
+        value: String,
+        maxAgeSeconds: Long,
+        httpOnly: Boolean = true,
+        secure: Boolean = true,
+    ): String {
+        val parts = mutableListOf("$name=$value", "Path=/", "Max-Age=$maxAgeSeconds", "SameSite=Lax")
+        if (httpOnly) parts.add("HttpOnly")
+        if (secure) parts.add("Secure")
+        return parts.joinToString("; ")
+    }
+}
+
+/**
+ * A set of public gRPC methods that do not require authentication.
+ */
+private val PUBLIC_METHODS =
+    setOf(
+        "snowballr.SnowballR/GetAuthenticationStatus",
+        "snowballr.SnowballR/Register",
+        "snowballr.SnowballR/Login",
+        "snowballr.SnowballR/RequestPasswordReset",
+        "snowballr.SnowballR/ResetPassword",
```

Comment:
We can use the method descriptors of the generated gRPC server code to get the full method names. This has the advantage that it is always up-to-date, and the code wouldn't compile if, e.g., a method wouldn't exist anymore.
```suggestion
        SnowballRGrpcKt.getAuthenticationStatusMethod.fullMethodName,
        SnowballRGrpcKt.registerMethod.fullMethodName,
        SnowballRGrpcKt.loginMethod.fullMethodName,
        SnowballRGrpcKt.requestPasswordResetMethod.fullMethodName,
        SnowballRGrpcKt.resetPasswordMethod.fullMethodName,
```

Reply:
Could you please replace `SERVICE_NAME` in `SnowballRServer.kt` with `SnowballRGrpcKt.serviceDescriptor.name`? I didn't find this back then, and this is the better solution.

Reply:
Great suggestion, I will change it!